### PR TITLE
Fixes #2040 - Prevent the room list from staying empty after cancelli…

### LINKED ISSUE #2040: Room list empty after cancelling search
Original report here https://github.com/vector-im/element-x-ios-rageshakes/issues/1018

```
Empty room list after a room search

To reproduce

start a room search
type a letter
cancel the search
The room list is empty NOK
```

### PR DESCRIPTION
**Pull request by stefanceriu**
…ng a search

### CODE REVIEW

File: ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift [hunk 1]
```diff
@@ -98,21 +98,19 @@ class HomeScreenViewModel: HomeScreenViewModelType, HomeScreenViewModelProtocol
...
-            .map(\.bindings.isSearchFieldFocused)
+            .map { SearchBarState(searchQuery: $0.bindings.searchQuery, isSearchFieldFocused: $0.bindings.isSearchFieldFocused) }
             .removeDuplicates()
-            .sink { [weak self] isSearchFieldFocused in
+            .sink { [weak self] _ in
                 guard let self else { return }
-                updateFilter(isSearchFieldFocused: isSearchFieldFocused, searchQuery: state.bindings.searchQuery)
+                // Don't capture the values here as combine behaves incorrectly and `isSearchFieldFocused` is sometimes
+                // turning to true after cancelling the search. Read them directly from the state in the updateFilter method instead.
+                updateFilter()
```
Comment (stefanceriu) [inline ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:113]: fwiw doug and I tried various combine solutions but it still wasn't behaving correctly, no idea what the underlying problem actually is. Sometimes `isSearchFieldFocused` just turns to true without anybody changing it.

Comment (alfogrillo) [reply ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:113]: I think the problem in the previous implementation was that the two combine chains were reading from `state.bindings` one value and receiving the other one from the publisher.
Did you try to read both from the publishers?
One idea can be this:
```swift
        let isSearchFieldFocused = context.$viewState.map(\.bindings.isSearchFieldFocused).removeDuplicates()
        let searchQuery = context.$viewState.map(\.bindings.searchQuery).removeDuplicates()
        isSearchFieldFocused
            .combineLatest(searchQuery)
            .sink { [weak self] isSearchFieldFocused, searchQuery in
                guard let self else { return }
                updateFilter(isSearchFieldFocused: isSearchFieldFocused, searchQuery: searchQuery)
            }
            .store(in: &cancellables)
```
If it works we could delete `struct SearchBarState: Equatable`

Comment (stefanceriu) [reply ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:113]: That what we figured as well but we did actually try this ^ and it was still wrong. Even the plain .map which gets the valure directly from bindings at every change was wrong at times. The only thing that did work was to fetch it from the state outside of the sink.

Comment (alfogrillo) [reply ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:113]: My memories on [objectwillchange](https://developer.apple.com/documentation/combine/observableobject/objectwillchange-2oa5v) are that published  values are always updated, but the corresponding values in the state may still be "old" if devs try to read them directly.
This is why I thought that reading both from a Combine chain is better.
So given the current implementation it's like inside `updateFilter()` we read the old values 🤔
Tbh I'm not sure why this works.
wdyt?

Comment (stefanceriu) [reply ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift:113]: Yeah, I know exactly what you mean, I have absolutely no idea either but empirically it does 😅

**Comment by alfogrillo** [comment]
LGTM!

**Review by alfogrillo — APPROVED** [review]

File: ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift (hunk 1 above)
Comment (stefanceriu) [reply ElementX/Sources/Screens/HomeScreen/HomeScreenViewModel.swift]: Talked about it offline, pushed another that uses the same approach but with an extra async dispatch and simpler implementation.

**Pull request merged**